### PR TITLE
Adds manual DM implementation of NEMOLite2D (for #7)

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
Build on my laptop fails with:

    mpif90 -Wall -Wsurprising -Wuninitialized -faggressive-function-elimination -Ofast -mtune=native -finline-limit=50000 -fopt-info-all=gnu_opt_report.txt -march=core2 -mtune=core2 -ffree-line-length-none  -fopenmp -I../../common -I/home/me/Projects/dl_esm_inf/finite_difference/src -I../../../../../shared/dl_timer/src -c time_step_mod.f90
    f951: Fatal Error: Reading module ‘model_mod’ at line 176 column 62: Expected left parenthesis

which is weird.

**Comment by arporter** [comment]
Having trouble getting everything working. If I fall back to e7b8292 and use b1347a2 of the haloex branch of dl_esm_inf then things are fine (i.e. before I tidied and brought branches up-to-date with master).

**Comment by arporter** [comment]
None of the commits since e7b8292 have updated any actual code (just meta-data and submodule stuff). Therefore, the problem must lie in my updates to dl_esm_inf.

**Comment by arporter** [comment]
Commit 9ed6b41 (stencil meta-data renaming) works with dl_esm_inf b808f8c ('fix build system after merge with master'). I am hitting problems moving between commits because I sometimes get the old link to parallel_utils_mod.f90 left lying around.

**Comment by arporter** [comment]
Move dl_esm_inf forward to 9aeabbc (fix waitall/waitany mistake) and things no longer work (although dl_esm_inf's tests still pass). Looks to be a problem with my re-factoring of dl_esm_inf.

**Comment by arporter** [comment]
Actually, the problem is that, by default, we do a serial build of dl_esm_inf and the `nemolite2d/common/Makefile` is building it with no target specified (and thus does the default build). If I enable debug output in the halo-swap code then I see no output - i.e. no halo swaps are performed.

**Comment by arporter** [comment]
This is now waiting on https://github.com/stfc/dl_esm_inf/pull/25.

**Comment by sergisiso** [comment]
Summary of the current status of this branch.

- NemoLite -> Original: Compiles and Runs
- NemoLite -> PSykal: Compiles and Runs (no perf improvement from OMP?)
- NemoLite -> manual_versions -> psykal_acc (ignoring pragmas): Compiles and run
- NemoLite -> manual_versions -> psykal_dm: Compiles and runs
- NemoLite -> manual_versions -> psykal_ocl: (I will fix this in the other branch)
- NemoLite -> manual_versions -> psykal_omp: Compiles and runs
- NemoLite -> manual_versions -> serial: Compiles and runs
- NemoLite -> manual_versions -> single_acc_file: (Unknown: asks for cudafor)

- Shallow -> SEQ: Compiles but run stops with: map_comms call needs to be implemented for periodic boundary conditions.
- Shallow -> OMP: Compiles and run
- Shallow -> Maxeler: (Unknown)

**Comment by sergisiso** [comment]
@arporter On Shallow -> SEQ the map_comms is called from dl_esm_inf grid_mod::init_grid line 467.

Should we just bypass this map_comms call if num_ranks == 1? or do we need to stub the call and set up some default values?

**Comment by arporter** [comment]
@sergisiso - I've had a look and I think we can probably just copy what `exchange_generic()` does and add:

   ! Do nothing if distributed-memory support is not enabled
    if(.not. DIST_MEM_ENABLED)return

at the beginning of the map_comms() routine.

**Comment by sergisiso** [comment]
@arporter With the updated dl_esm_inf now all version of nemolite and shallow compile and run. I have now been looking at the checksums and times (using the same namelist everywhere) and I found:

Nemolite:
psykal_serial
ua checksum =   0.71987034E+04
va checksum =   0.15527224E+05
time: 0.38610E+02

psykal_omp 4t
ua checksum =   0.71987034E+04
va checksum =   0.15527224E+05
time: 0.10990E+02

psykal_dm 1 MPI
ua checksum =   0.75086026E+04
va checksum =   0.15507322E+05
~60 seconds

psykal_dm 4 MPI
ua checksum =   0.00000000E+00
va checksum =   0.00000000E+00
ua checksum =   0.40428062E+04
va checksum =   0.84314695E+04
ua checksum =   0.40415614E+04
va checksum =   0.84314632E+04
ua checksum =   0.00000000E+00
va checksum =   0.00000000E+00

Shallow:
./shallow_base.exe
psi initial CHECKSUM =   0.6641518585171037E+10
P initial CHECKSUM =   0.8192000000000004E+09
U initial CHECKSUM =   0.3258838709766664E+04
V initial CHECKSUM =   0.3258838709766664E+04
P CHECKSUM after   2000 steps =   0.8191999804318038E+09
U CHECKSUM after   2000 steps =   0.3258876676141899E+04
V CHECKSUM after   2000 steps =   0.3258798989007143E+04

./shallow_gen.exe
psi initial CHECKSUM =   0.6641518585171037E+10
P initial CHECKSUM =   0.8192000000000004E+09
U initial CHECKSUM =   0.3258838709766664E+04
V initial CHECKSUM =   0.3258838709766664E+04
P CHECKSUM after   2000 steps =   0.8064499847728990E+09
U CHECKSUM after   2000 steps =   0.1696068206302023E+06
V CHECKSUM after   2000 steps =   0.1675531126992660E+06

The issues are:
- MPI checksums are only of partial arrays right? So they should not be the same that other runs, but the nemolite 4 MPI has 2 processes with 0 checksums which seems wrong.
- There is a performance degradation btw psykal_serial and psykal_dm 1 MPI.
- Shallow U and V checksums after 2000it are quite different.

Do you know why some of these happen? Should we fix them in this PR or add them in the issue tracker?

**Comment by arporter** [comment]
You're being very thorough - thank you! You're right that the MPI sums will be partial. Depending on the domain size, it is possible that any waves will not have reached sufficiently far into the domain to affect those two processors. Might be worth checking on a smaller model domain. (Or you can visualise the output using e.g. Gnuplot and check that everything is OK. I had got it to a point where everything was working but that was several commits ago now.)
As long as the DM version is producing ~correct results (according to visualisation) then I suggest we proceed with this PR as it will get us to a consistent state. As you suggest, we can open issues for the other problems.

File: ocean/nemo/nemolite2d/manual_versions/psykal_dm/time_step_mod.f90 [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    type(r2d_field), intent(inout) :: ua, va, ssha, ssha_u, ssha_v
+    type(r2d_field), intent(in)    :: hu, hv, ht
+    ! Locals
+    integer :: it, ji, jj
+    integer :: idepth = 1
+
+    ! We don't know the state of a field's halos on entry to this
+    ! invoke so we have to check...
+    ! (hu, hv and ht are constant so no need to worry about them)
+!!$    if(sshn_u%is_dirty(depth=idepth))then
```
Comment (sergisiso) [inline ocean/nemo/nemolite2d/manual_versions/psykal_dm/time_step_mod.f90]: Are this checks not needed anymore? In this case remove and update the comment.

File: ocean/nemo/nemolite2d/manual_versions/psykal_dm/nemolite2d.f90 [hunk 2]
```diff
@@ -120,15 +109,16 @@ program gocean2d
...
   call model_write_log("((A))", 'Simulation finished!!')
 
+  call gocean_finalise()
+
 end program gocean2d
 
 !+++++++++++++++++++++++++++++++++++
 
-subroutine step(istp,           &
+subroutine step(grid, istp, &
```
Comment (sergisiso) [inline ocean/nemo/nemolite2d/manual_versions/psykal_dm/nemolite2d.f90]: Remove commented invoke at the end of this subroutine

**Comment by arporter** [comment]
Proof of the fact that on 01/05/2019 at 09:29am the halo-swapping was all working correctly:

splash

File: ocean/nemo/nemolite2d/manual_versions/psykal_dm/time_step_mod.f90 (hunk 1 above)
Comment (arporter) [reply ocean/nemo/nemolite2d/manual_versions/psykal_dm/time_step_mod.f90]: They aren't strictly needed in this manual version and dl_esm_inf doesn't yet have the necessary support. I've removed the commented-out code and updated the comment with a reference to the Issue in dl_esm_inf.

File: ocean/nemo/nemolite2d/manual_versions/psykal_dm/nemolite2d.f90 (hunk 2 above)
Comment (arporter) [reply ocean/nemo/nemolite2d/manual_versions/psykal_dm/nemolite2d.f90]: Done.

**Comment by arporter** [comment]
I think I've now addressed all comments (and created a new Issue for the MPI-related problems).
Ready for next review.

**Comment by sergisiso** [comment]
@arporter Everything looks right, I will merge this branch now

**Pull request merged**